lib/resty/upstream/healthcheck.lua

```diff
@@ -209,9 +241,41 @@ local function peer_error(ctx, is_backup, id, peer, ...)
   peer_fail(ctx, is_backup, id, peer)
 end
 
-local function check_peer(ctx, id, peer, is_backup)
+local function copy_status_if_checked(peer, is_backup, ctx)
+    local dict = ctx.dict
+    local u = ctx.upstream
+
+    local id = peer.id
+    local set_in_current_upstream = dict:get(gen_peer_key("ok:", u, is_backup, id))
+
+    local checked_peer_upstream = dict:get('checked_peers_upstream:' .. peer.name)
+    local peer_set_in_different_upstream = dict:get(gen_peer_key("ok:", checked_peer_upstream, is_backup, id))
+
+    if set_in_current_upstream or peer_set_in_different_upstream then
+        return peer_ok(ctx, is_backup, id, peer)
+    end
+
+    local noks, _ = dict:get(gen_peer_key("nok:", checked_peer_upstream, is_backup, id))
+    if noks then
+        peer_fail(ctx, is_backup, id, peer)
+    end
+end
+
+local function propogate_peer_status(ctx)
+    local ppeers, bpeers = get_all_peers(ctx.upstream)
```

Comment:
During initialization we fetch peers into `ctx` table. I think we should use them here instead of refetching peers:
```
local ppeers, bpeers = ctx.primary_peers, ctx.backup_peers
```